### ISSUE TITLE
Tab Complete not working in GP 6.0.0?

### ISSUE THREAD

**Issue by KAdamsInCO** [greenplum-db/gpdb#8575]
### Greenplum version or build

GPDB 6.0.0 (GA)

gpadmin=# select version();
                                                                                               version                                                                                                
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 9.4.24 (Greenplum Database 6.0.0 build commit:a2ff2fc318c327d702e66ab58ae4aff34c42296c) on x86_64-unknown-linux-gnu, compiled by gcc (GCC) 6.4.0, 64-bit compiled on Aug 28 2019 21:29:36
(1 row)

### OS version and uname -a

[gpadmin@mdw ~]$ uname -a
Linux mdw 3.10.0-957.27.2.el7.x86_64 #1 SMP Mon Jul 29 17:46:05 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

[gpadmin@mdw ~]$ cat /etc/redhat-release 
CentOS Linux release 7.6.1810 (Core) 

### Expected behavior

Tab complete should work for DB objects within psql.

### Actual behavior

Tab complete is not working in psql. It works for all UPPERCASE keywords, but not lowercase keywords. Is there something I didn't do right when initializing this cluster? I set this up just like the GP 5.x clusters I normally work with.

### Step to reproduce the behavior

Start something like a SQL statement, then try to tab-complete the table name. Hitting the TAB key just repeats the command over and over:

[gpadmin@mdw ~]$ psql
psql (9.4.24)
Type "help" for help.

gpadmin=# select * from gp_seg \<tab\>

gpadmin=# select * from gp_seg \<tab\>

gpadmin=# select * from gp_seg 

Although, tab complete does work for all upper-case words:

So typing CHECK\<tab\>:

gpadmin=# CHECK\<tab\>

Produces this:

gpadmin=# CHECKPOINT

**Comment by gkokolatos** [comment]
Thank you for reporting the issue @KAdamsInCO 

I took a look. I have a suspicion with regards to the cause yet I am afraid I am suffering from a `cannot reproduce` syndrome.

It would be helpfull if you could provide the locale settings of your environment where you run `psql` from, example by running `locale -a`. Also if you can provide the enconding of the session and the server, example by issuing `show client_encoding;` and `show lc_ctype;` in your session, it would be great.

Hopefully this info would help me reprodruce.

**Comment by gkokolatos** [comment]
Thank you again @KAdamsInCO for the additional information.

I still suffer from a `can not reproduce` which is less than ideal. I apologise. 

I have tried to reproduce with a variety of locales, both on Ubuntu 18.04.3 LTS and CentOS 7. Those attempts also include distinct readline versions from 6.2 to 7.

Either I am missing something obvious, which I don't discard, or I am not setting up my environment properly for reproduction.

For visibility, my current focus has been in the readline library installed and the locale used. The interactive terminal is using the input parsed into readline's `rl_line_buffer`. It tokenises the input using a simple needle in haystack on `"\t\n@$><=;|&{() "`. Then it tries to match the tokens with some hardcoded sequences in order to find a rule to apply. The matching is done via `pg_strcasecmp()` invocations which can suffer in some of the less used locales.

My hypothesis has been that readline returned a buffer that either `strchr()`(unlikely) or `pg_strcasecmp()` (more likely) can not handle gracefully. I have not been able to prove my hypothesis.

I will ask once more for your patience and some additional information:

1)  Are you working on a vanilla CentOS Linux release 7.6.1810 (Core)? If not and you have some reproducible environment, please point me to it.

2) Which flags do you use to compile with?

Thanks again!

**Comment by mark-a-s** [comment]
I've got this problem too FWIW, with the rpm pulled from git.



gpadmin=# show client_encoding;
 client_encoding
-----------------
 UTF8
(1 row)

gpadmin=# show lc_ctype ;
 lc_ctype
----------
 C
(1 row)

**Comment by Eulerizeit** [comment]
Potential thing to look at. 

I was talking to someone smarter than me and they wondered if it might be related to libedit/libreadline and what we are compiling against. I’m not sure what we actually dynamically link against and which one is currently set in for the configure flags. We don’t distribute either anymore.

**Comment by KAdamsInCO** [comment]
Another GPDB user having the same issue.

Thanks,

Keaton




From: Mark Sloan <Mark.Sloan@clickfox.com>
Date: Friday, September 13, 2019 at 1:19 PM
To: Keaton Adams <kadams@pivotal.io>
Subject: Re: [greenplum-db/gpdb] Tab Complete not working in GP 6.0.0? (#8575)

I have this problem on regular old centos 7.


Easy examples: I expect \c to tab complete to the list of databases, not the key syntax words.



[gpadmin@gpdb-sn ~]$ psql

psql (9.4.24)

Type "help" for help.



gpadmin=# \c

ABORT          CREATE         FETCH          REFRESH        SHOW

ALTER          DEALLOCATE     GRANT          REINDEX        START

ANALYZE        DECLARE        INSERT         RELEASE        TABLE

BEGIN          DELETE FROM    LISTEN         RESET          TRUNCATE

CHECKPOINT     DISCARD        LOAD           REVOKE         UNLISTEN

CLOSE          DO             LOCK           ROLLBACK       UPDATE

CLUSTER        DROP           MOVE           SAVEPOINT      VACUUM

COMMENT        END            NOTIFY         SECURITY LABEL VALUES

COMMIT         EXECUTE        PREPARE        SELECT         WITH

COPY           EXPLAIN        REASSIGN       SET

gpadmin=# \l

                                             List of databases

         Name         |      Owner      | Encoding | Collate | Ctype |

Access privileges

----------------------+-----------------+----------+---------+-------+----------

---------------------------

 direct_attribute_poc | gpadmin         | UTF8     | C       | C     |

 dolphin_demo         | cf_dolphin_demo | UTF8     | C       | C     | cf_dolphi

n_demo=CTc/cf_dolphin_demo+

                      |                 |          |         |       | =Tc/cf_do

lphin_demo

 gpadmin              | gpadmin         | UTF8     | C       | C     |

 jsonb_poc            | gpadmin         | UTF8     | C       | C     |

 marks_dolphin        | gpadmin         | UTF8     | C       | C     |

 marks_test           | gpadmin         | UTF8     | C       | C     |

 postgres             | gpadmin         | UTF8     | C       | C     |

 rchundi              | gpadmin         | UTF8     | C       | C     |

 temp_msloan          | gpadmin         | UTF8     | C       | C     |

 template0            | gpadmin         | UTF8     | C       | C     | =c/gpadmi

n                         +

                      |                 |          |         |       | gpadmin=C

Tc/gpadmin

 template1            | gpadmin         | UTF8     | C       | C     | gpadmin=C

Tc/gpadmin                +

                      |                 |          |         |       | =c/gpadmi

n

 tjdb                 | cf_tjdb         | UTF8     | C       | C     | cf_tjdb=C

Tc/cf_tjdb                +

                      |                 |          |         |       | =Tc/cf_tj

db

(12 rows)



gpadmin=# \q

[gpadmin@gpdb-sn ~]$ cat /etc/redhat-release

CentOS Linux release 7.6.1810 (Core)

[gpadmin@gpdb-sn ~]$

[gpadmin@gpdb-sn ~]$ env

XDG_SESSION_ID=3746

HOSTNAME=gpdb-sn.clickfox.net

MASTER_DATA_DIRECTORY=/data/master/gpseg-1

TERM=xterm-256color

SHELL=/bin/bash

HISTSIZE=1000

SSH_CLIENT=192.168.12.126 51324 22

GPHOME=/usr/local/greenplum-db/.

SSH_TTY=/dev/pts/0

USER=gpadmin

LD_LIBRARY_PATH=/usr/local/greenplum-db/./lib:/usr/local/greenplum-db/./ext/python/lib:

LS_COLORS=rs=0:di=38;5;27:ln=38;5;51:mh=44;38;5;15:pi=40;38;5;11:so=38;5;13:do=38;5;5:bd=48;5;232;38;5;11:cd=48;5;232;38;5;3:or=48;5;232;38;5;9:mi=05;48;5;232;38;5;15:su=48;5;196;38;5;15:sg=48;5;11;38;5;16:ca=48;5;196;38;5;226:tw=48;5;10;38;5;16:ow=48;5;10;38;5;21:st=48;5;21;38;5;15:ex=38;5;34:*.tar=38;5;9:*.tgz=38;5;9:*.arc=38;5;9:*.arj=38;5;9:*.taz=38;5;9:*.lha=38;5;9:*.lz4=38;5;9:*.lzh=38;5;9:*.lzma=38;5;9:*.tlz=38;5;9:*.txz=38;5;9:*.tzo=38;5;9:*.t7z=38;5;9:*.zip=38;5;9:*.z=38;5;9:*.Z=38;5;9:*.dz=38;5;9:*.gz=38;5;9:*.lrz=38;5;9:*.lz=38;5;9:*.lzo=38;5;9:*.xz=38;5;9:*.bz2=38;5;9:*.bz=38;5;9:*.tbz=38;5;9:*.tbz2=38;5;9:*.tz=38;5;9:*.deb=38;5;9:*.rpm=38;5;9:*.jar=38;5;9:*.war=38;5;9:*.ear=38;5;9:*.sar=38;5;9:*.rar=38;5;9:*.alz=38;5;9:*.ace=38;5;9:*.zoo=38;5;9:*.cpio=38;5;9:*.7z=38;5;9:*.rz=38;5;9:*.cab=38;5;9:*.jpg=38;5;13:*.jpeg=38;5;13:*.gif=38;5;13:*.bmp=38;5;13:*.pbm=38;5;13:*.pgm=38;5;13:*.ppm=38;5;13:*.tga=38;5;13:*.xbm=38;5;13:*.xpm=38;5;13:*.tif=38;5;13:*.tiff=38;5;13:*.png=38;5;13:*.svg=38;5;13:*.svgz=38;5;13:*.mng=38;5;13:*.pcx=38;5;13:*.mov=38;5;13:*.mpg=38;5;13:*.mpeg=38;5;13:*.m2v=38;5;13:*.mkv=38;5;13:*.webm=38;5;13:*.ogm=38;5;13:*.mp4=38;5;13:*.m4v=38;5;13:*.mp4v=38;5;13:*.vob=38;5;13:*.qt=38;5;13:*.nuv=38;5;13:*.wmv=38;5;13:*.asf=38;5;13:*.rm=38;5;13:*.rmvb=38;5;13:*.flc=38;5;13:*.avi=38;5;13:*.fli=38;5;13:*.flv=38;5;13:*.gl=38;5;13:*.dl=38;5;13:*.xcf=38;5;13:*.xwd=38;5;13:*.yuv=38;5;13:*.cgm=38;5;13:*.emf=38;5;13:*.axv=38;5;13:*.anx=38;5;13:*.ogv=38;5;13:*.ogx=38;5;13:*.aac=38;5;45:*.au=38;5;45:*.flac=38;5;45:*.mid=38;5;45:*.midi=38;5;45:*.mka=38;5;45:*.mp3=38;5;45:*.mpc=38;5;45:*.ogg=38;5;45:*.ra=38;5;45:*.wav=38;5;45:*.axa=38;5;45:*.oga=38;5;45:*.spx=38;5;45:*.xspf=38;5;45:

SSH_AUTH_SOCK=/tmp/ssh-1NYwkEkkfn/agent.15478

MAIL=/var/spool/mail/gpadmin

PATH=/usr/local/greenplum-db/./bin:/usr/local/greenplum-db/./ext/python/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/gpadmin/.local/bin:/home/gpadmin/bin

PWD=/home/gpadmin

LANG=en_US.UTF-8

PYTHONHOME=/usr/local/greenplum-db/./ext/python

HISTCONTROL=ignoredups

SHLVL=1

HOME=/home/gpadmin

PYTHONPATH=/usr/local/greenplum-db/./lib/python

LOGNAME=gpadmin

SSH_CONNECTION=192.168.12.126 51324 10.101.1.90 22

LESSOPEN=||/usr/bin/lesspipe.sh %s

XDG_RUNTIME_DIR=/run/user/1001

_=/usr/bin/env



SELECT count(*) FROM <tab><tab> doen’t give me a list of tables.

If you do select count(*) FROM start_of_table_name<tab> it won’t complet, and start_of_table_name<tab><tab> won’t list possilbes if it can’t complete it.

SELECT … FROM table WHERE <tab><tab> should list columns.




Mark Sloan
DevOps Engineer
P 303.951.3490,,5432

**Comment by gkokolatos** [comment]
> Potential thing to look at.
> 
> I was talking to someone smarter than me and they wondered if it might be related to libedit/libreadline and what we are compiling against. I’m not sure what we actually dynamically link against and which one is currently set in for the configure flags. We don’t distribute either anymore.

I agree with the person you were talking. It was my first assumption and I tried to reproduce with distinct edit/readline versions as it is mentioned in my response above to no avail.

However @mark-a-s points out that rpm's are used whereas I have been building from source code. I will try with the rpm's on vanilla centos 7.

Apologies for the long silence, I am still down with the flu and the headache is still not permissive of work. I hope to be able to take a look again tomorrow.

**Comment by gkokolatos** [comment]
The issue is indeed reproducible when installing the rpm packages in a vanilla centos 7.

The root cause of the problem is indeed related to libedit and the libraries the binary is linked with.
For visibility, the shipped psql binary is linked against libedit alone.

```
 0x0000000000000001 (NEEDED)             Shared library: [libpq.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libssl.so.10]
 0x0000000000000001 (NEEDED)             Shared library: [libedit.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```

 Libedit in centos 7 `libedit.so.0.0.42` (the only version I checked although I am confident that it might affect earlier versions) is providing the `rl_*` and related symbols that the code is expecting, yet they are not necessarily the same.

```
(gdb) p /x rl_readline_version
$1 = 0x402
```

and

```
    69: 0000000000674a90     8 OBJECT  GLOBAL DEFAULT   25 rl_basic_word_break_chara
    70: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND using_history
    71: 0000000000674a98     8 OBJECT  GLOBAL DEFAULT   25 rl_line_buffer
```

In this specific case, `rl_line_buffer` is not populated with the contents of the interactive terminal, but with the last line of the written/flushed history. Which means that in the first interactive session a user will launch, `rl_line_buffer` will be empty and no tab completion information will be able to be launched. The `text` argument of the tab completion function `psql_completion` will be populated though which will lead the code to add ' ' when a statement like `SELECT * FROM gp_` is autocompleted upon.


This means that if a statement like `SELECT * FROM gp_segment_configuration;` is launched in an interactive session, then the session is closed and a new session starts, all autocompletion in the current session will be based as if the contents of the autocompleted statement where `SELECT * FROM` regardless of the current statement.

Example:
```
Breakpoint 1, psql_completion (text=0x11f7d90 "gp_s", start=14, end=18) at tab-complete.c:909
(gdb) p rl_line_buffer
$3 = 0x11ea620 "select pg_backend_pid();\n"
```

One short term solution until an amended package is released, is to use a Greenplum psql binary compiled from source.

All of the above do not yet explain the uppercase/lowercase part of the issue though for the moment I am willing to dismiss it as a side effect of the above. 

Thank you for raising the issue.

**Comment by skahler-pivotal** [comment]
Just did an rpm install and can verify the same issue

**Comment by gkokolatos** [comment]
> Just did an rpm install and can verify the same issue

Great! Consistency is key. There is a PR open [/pull/8683] to address the issue. New rpm's will get baked shortly.

**Comment by bradfordboyle** [comment]
> Just did an rpm install and can verify the same issue

Where did you get the RPM from? Was it from https://github.com/greenplum-db/gpdb/releases/tag/6.0.0?

**Comment by KAdamsInCO** [comment]
Yes, correct.

**Comment by gaos1** [comment]
fixed in master and 6.0.1.
